mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,370 @@
+                    FuncArg2BlockArg[operand] = p;
+                    p++;
+                }
+            }
+
+            // If an op is needed by post processing, it should be returned
+            if (std::find(NecessaryOpsForPostProcessing.begin(),
+                          NecessaryOpsForPostProcessing.end(),
+                          op) != NecessaryOpsForPostProcessing.end()) {
+                RetOps.push_back(op);
+                for (size_t i = 0; i < op->getNumResults(); i++) {
+                    RetOpsTypes.push_back(op->getResultTypes()[i]);
+                }
+            }
+        }
+
+        // We process the tapes in reverse order
+        // Hence from the viewpoint of the earlier tapes,
+        // a processed later tape (which has become a call op) is also its "post processing"
+        // The later tape's arguments need to be returned by the earlier tapes
+        for (auto op : ArgOps) {
+            NecessaryOpsForPostProcessing.push_back(op);
+        }
+
+        FunctionType fType = FunctionType::get(module->getContext(), ArgOpsTypes, RetOpsTypes);
+
+        // Create the tape function.
+        // Keep the original function's attributes
+        // We do not want the name and the type of the original multitape function:
+        // The type could be different, and we want different names for each tape
+        SmallVector<NamedAttribute> FuncAttrs;
+        for (auto attr : FullOriginalFuncAttrs) {
+            StringRef attrname = attr.getName();
+            if ((attrname != "sym_name") && (attrname != "function_type")) {
+                FuncAttrs.push_back(attr);
+            }
+        }
+
+        // The arguments don't need to carry attributes
+        builder.setInsertionPointAfter(OriginalMultitapeFunc);
+        func::FuncOp TapeFunc = builder.create<func::FuncOp>(
+            module->getLoc(),
+            OriginalMultitapeFuncName.str() + "_tape_" + std::to_string(tape_number), fType,
+            FuncAttrs);
+
+        // Call the newly made tape function in the original function
+        // The i-th argument corresponds to the i-th op in Argops
+        // The callop results should replace the retops in their users
+
+        builder.setInsertionPoint(TapeOps->front());
+        func::CallOp callOp = builder.create<func::CallOp>(module->getLoc(), TapeFunc, CallArgs);
+
+        // Inject the tape's operations into the tape function
+        Block *entrybb = TapeFunc.addEntryBlock();
+        builder.setInsertionPointToEnd(entrybb);
+
+        //<< entrybb->getArgument(0) << entrybb->getArgument(1) << "\n";
+
+        std::map<Operation *, Operation *> ClonedOpShadows;
+        SmallVector<Value> ClonedRetValues;
+        for (auto op : *TapeOps) {
+            ClonedOpShadows[op] = op->clone();
+            builder.insert(ClonedOpShadows[op]);
+        }
+
+        // for (auto op : *TapeOps) {
+        //     op->replaceAllUsesWith(ClonedOpShadows[op]);
+        // }
```

Comment:
This native replacement has issues, as it replaces uses outside the cloned shadows, and we do not want that. The paragraph below handles the shadow's replacement properly.

Turn this into a proper comment.